pglookout/webserver.py

```diff
@@ -17,43 +22,62 @@
-    cluster_state = None
-    log = None
-    cluster_monitor_check_queue = None
-    allow_reuse_address = True
+    allow_reuse_address: bool = True
+
+    def __init__(
+        self,
+        address: str,
+        port: int,
+        RequestHandlerClass: type[BaseHTTPRequestHandler],
+        cluster_state: dict[str, MemberState],
+        log: Logger,
+        cluster_monitor_check_queue: Queue[str],
+    ) -> None:
+        super().__init__((address, port), RequestHandlerClass)
+        self.cluster_state: dict[str, MemberState] = cluster_state
+        self.log: Logger = log
+        self.cluster_monitor_check_queue: Queue[str] = cluster_monitor_check_queue
 
 
 class WebServer(Thread):
-    def __init__(self, config, cluster_state, cluster_monitor_check_queue):
-        Thread.__init__(self)
-        self.config = config
-        self.cluster_state = cluster_state
-        self.cluster_monitor_check_queue = cluster_monitor_check_queue
-        self.log = getLogger("WebServer")
-        self.address = self.config.get("http_address", "")
-        self.port = self.config.get("http_port", 15000)
-        self.server = None
+    def __init__(
+        self, config: Config, cluster_state: dict[str, MemberState], cluster_monitor_check_queue: Queue[str]
+    ) -> None:
+        super().__init__()
+        self.config: Config = config
+        self.cluster_state: dict[str, MemberState] = cluster_state
+        self.cluster_monitor_check_queue: Queue[str] = cluster_monitor_check_queue
+        self.log: Logger = getLogger("WebServer")
+        self.address: str = self.config.get("http_address", "")
+        self.port: int = self.config.get("http_port", 15000)
+        self.server: ThreadedWebServer | None = None
         self.log.debug("WebServer initialized with address: %r port: %r", self.address, self.port)
-        self.is_initialized = threading.Event()
+        self.is_initialized: threading.Event = threading.Event()
 
-    def run(self):
+    def run(self) -> None:
         # We bind the port only when we start running
-        self.server = ThreadedWebServer((self.address, self.port), RequestHandler)
-        self.server.cluster_state = self.cluster_state
-        self.server.log = self.log
-        self.server.cluster_monitor_check_queue = self.cluster_monitor_check_queue
+        self.server = ThreadedWebServer(
+            address=self.address,
+            port=self.port,
+            RequestHandlerClass=RequestHandler,
+            cluster_state=self.cluster_state,
+            log=self.log,
+            cluster_monitor_check_queue=self.cluster_monitor_check_queue,
+        )
         self.is_initialized.set()
         self.server.serve_forever()
 
-    def close(self):
-        if self.server:
-            self.log.debug("Closing WebServer")
-            self.server.shutdown()
-            self.log.debug("Closed WebServer")
+    def close(self) -> None:
+        if self.server is None:
```

Comment:
did you check if server classes  override ```__bool__``` ?

Reply:
Yes, I digged into the whole MRO.
Also, with type hints, it appeared that this is `ThreadedWebServer | None`, is initialized to `None`, and it only set in the `run()`. This is a safe guard to make sure that the `close()` is only called after `run()`. The condition is necessary so that `self.server` has a `.shutdown` attr, and we won't raise an `AttributeError`.

If at a point in the future someone implements a `__bool__`, we should still ignore that and only care about whether it's not `None`, to have the attr defined.